packages/fxa-auth-server/scripts/paypal-refund-fixer.ts

```diff
@@ -0,0 +1,182 @@
+    private paypalHelper: PayPalHelper
+  ) {
+    this.stripe = (this.stripeHelper as any).stripe as stripe;
+  }
+
+  private activeSubscriptions(
+    subscriptions: stripe.Subscription[] | undefined
+  ) {
+    if (!subscriptions) {
+      return [];
+    }
+    return subscriptions.filter((sub) =>
+      ACTIVE_SUBSCRIPTION_STATUSES.includes(sub.status)
+    );
+  }
+
+  private async *nonRefundedPaypalInvoices(limit?: number) {
+    const startDate = new Date(2021, 3, 20).getTime() / 1000;
+    let count = 0;
+    for await (const invoice of this.stripe.invoices.list({
+      limit: 100,
+      status: 'paid',
+      collection_method: 'send_invoice',
+      created: { gt: startDate },
+    })) {
+      if (
+        !invoice.metadata![STRIPE_INVOICE_METADATA.PAYPAL_TRANSACTION_ID] ||
+        invoice.metadata![
+          STRIPE_INVOICE_METADATA.PAYPAL_REFUND_TRANSACTION_ID
+        ] ||
+        invoice.post_payment_credit_notes_amount === 0
+      ) {
+        continue;
+      }
+      yield invoice;
+      count++;
+      if (limit && count >= limit) {
+        break;
+      }
+    }
+  }
+
+  async zeroAccountBalance(customer: stripe.Customer) {
+    if (customer.balance >= 0) {
+      return;
+    }
+    const zeroAmount = Math.abs(customer.balance);
+    await this.stripe.customers.createBalanceTransaction(customer.id, {
+      amount: zeroAmount,
+      currency: customer.currency!,
+    });
+  }
+
+  async issueRefund(invoice: stripe.Invoice) {
+    const transactionId = this.stripeHelper.getInvoicePaypalTransactionId(
+      invoice
+    );
+    if (!transactionId) {
+      return;
+    }
+    const refundResponse = await this.paypalHelper.refundTransaction({
+      idempotencyKey: invoice.id,
+      transactionId: transactionId,
+    });
+    const success = ['instant', 'delayed'];
+    if (success.includes(refundResponse.refundStatus.toLowerCase())) {
+      await this.stripeHelper.updateInvoiceWithPaypalRefundTransactionId(
+        invoice,
+        refundResponse.refundTransactionId
+      );
+    } else {
+      this.log.error('issueRefund', {
+        message: 'PayPal refund transaction unsuccessful',
+        invoiceId: invoice.id,
+        transactionId,
+        refundResponse,
+      });
+    }
+  }
+
+  async fixAffectedCustomers() {
+    let i = 0;
+    for await (const invoice of this.nonRefundedPaypalInvoices()) {
+      const customer = await this.stripe.customers.retrieve(
+        invoice.customer as string,
+        { expand: ['subscriptions'] }
+      );
+      const subs = customer.deleted
+        ? []
+        : this.activeSubscriptions(customer.subscriptions?.data);
+      this.log.info('Invoice found', {
+        invoiceId: invoice.id,
+        customerId: customer.id,
+        balance: customer.deleted ? 0 : customer.balance,
+        subs: subs.length,
+        deleted: !!customer.deleted,
+      });
+      i++;
+
+      if (!customer.deleted) {
+        await this.zeroAccountBalance(customer);
```

Comment:
Should this come after `issueRefund`, in case that fails? tbh i don't know enough context to know which should go first